src/App.tsx

```diff
@@ -2,24 +1,66 @@
 import './App.css';
-import { getNumbers } from './utils';
 
-// eslint-disable-next-line @typescript-eslint/no-unused-vars
+import React, { useEffect, useState } from 'react';
+import { BrowserRouter, Route, Routes, useNavigate } from 'react-router-dom';
+
+import { getNumbers, maybeParseInt } from './utils';
+
+import { useUrlQuery } from './hooks/useUrlQuery';
+
+import { Pagination } from './components/Pagination/Pagination';
+
 const items = getNumbers(1, 42).map(n => `Item ${n}`);
 
-export const App: React.FC = () => {
+const RoutedApp: React.FC = () => {
+  const query = useUrlQuery();
+  const navigate = useNavigate();
+
+  const [perPage, setPerPage] = useState(
+    maybeParseInt(query.get('perPage')) ?? 5,
+  );
+  const [currentPage, setCurrentPage] = useState(
+    maybeParseInt(query.get('page')) ?? 1,
+  );
+
+  useEffect(() => {
+    if (!query.get('page') || !query.get('perPage')) {
+      navigate(`?page=${currentPage}&perPage=${perPage}`);
+    }
+  }, [perPage, query, navigate, currentPage]);
+
+  const handlePageChange = (newPage: number) => {
+    setCurrentPage(newPage);
+    navigate(`?page=${newPage}&perPage=${perPage}`);
+  };
+
+  const handlePerPageChange = (newPerPage: number) => {
+    setPerPage(newPerPage);
+    setCurrentPage(1);
+
+    navigate(`?page=${currentPage}&perPage=${newPerPage}`);
+  };
+
+  const firstItemOnPage = (currentPage - 1) * perPage + 1;
+  const lastItemOnPage = Math.min(currentPage * perPage, items.length);
+
   return (
     <div className="container">
       <h1>Items with Pagination</h1>
 
       <p className="lead" data-cy="info">
-        Page 1 (items 1 - 5 of 42)
+        Page {currentPage} (items {firstItemOnPage} - {lastItemOnPage} of{' '}
+        {items.length})
       </p>
 
       <div className="form-group row">
         <div className="col-3 col-sm-2 col-xl-1">
           <select
             data-cy="perPageSelector"
             id="perPageSelector"
-            className="form-control">
+            className="form-control"
+            defaultValue="5"
+            onChange={event => handlePerPageChange(Number(event.target.value))}
```

Comment:
The `defaultValue` attribute on the select element is set to '5', but this value should be derived from the state or URL query parameters to remain consistent with the current selection. Otherwise, the default value in the select element might not match the actual page size being used.